File: vuln-sandbox/node/taint/ssrf-2hop.js
// Taint: SSRF — 2-hop (JS)
// Source: req.body.endpoint → intermediate: fetchData() → Sink: axios.get
// Expected: TruePositive (taint/cwe918)
const express = require('express');
const axios = require('axios');
const app = express();

async function fetchData(endpoint) {
    return axios.get(endpoint);                 // SINK: outbound HTTP request
}

app.post('/fetch', async (req, res) => {
    const endpoint = req.body.endpoint;         // SOURCE: HTTP request body
    const data = await fetchData(endpoint);     // INTERMEDIATE
    res.json(data);
});

File: vuln-sandbox/python/taint/ssrf-2hop.py
# Taint: SSRF — 2-hop (Python)
# Source: request.args.get('endpoint') → intermediate: fetch() → Sink: requests.post
# Expected: TruePositive (taint/cwe918)
from flask import Flask, request
import requests

app = Flask(__name__)

def fetch(endpoint):
    return requests.post(endpoint, json={})    # SINK: outbound HTTP request

@app.route('/fetch')
def do_fetch():
    endpoint = request.args.get('endpoint')    # SOURCE: HTTP request param
    return fetch(endpoint).text                # INTERMEDIATE
